File: frontend/src/app/services/category.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/api-response.model';
import { Category, CategoryRequest } from '../models/category.model';

@Injectable({ providedIn: 'root' })
export class CategoryService {
  private http = inject(HttpClient);
  private readonly API = '/api/categories';

  getAll(): Observable<ApiResponse<Category[]>> {
    return this.http.get<ApiResponse<Category[]>>(this.API);
  }

  getById(id: number): Observable<ApiResponse<Category>> {
    return this.http.get<ApiResponse<Category>>(`${this.API}/${id}`);
  }

  create(req: CategoryRequest): Observable<ApiResponse<Category>> {
    return this.http.post<ApiResponse<Category>>(this.API, req);
  }

  update(id: number, req: CategoryRequest): Observable<ApiResponse<Category>> {
    return this.http.put<ApiResponse<Category>>(`${this.API}/${id}`, req);
  }

  delete(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API}/${id}`);
  }
}

File: backend/src/main/java/com/craftcorner/module/category/service/CategoryService.java
package com.craftcorner.module.category.service;

import com.craftcorner.module.category.dto.CategoryDto;
import com.craftcorner.module.category.dto.CategoryRequest;

import java.util.List;

public interface CategoryService {
    CategoryDto create(CategoryRequest request);
    List<CategoryDto> getAll();
    CategoryDto getById(Long id);
    CategoryDto update(Long id, CategoryRequest request);
    void delete(Long id);
    void toggleActive(Long id);
}
